source/non_matching/coupling.cc

```diff
@@ -330,22 +370,51 @@ namespace NonMatching
     const unsigned int n_q_points = quad.size();
     const unsigned int n_active_c =
       immersed_dh.get_triangulation().n_active_cells();
-    std::vector<Point<spacedim>> all_points(n_active_c * n_q_points);
-
-    // Collecting all points
+    std::vector<Point<spacedim>> all_points;
+    // Keep track of which cells we actually used
+    std::vector<unsigned int> used_cells;
     {
-      unsigned int c = 0;
+      FEValues<dim1, spacedim> fe_v(immersed_mapping,
+                                    immersed_fe,
+                                    quad,
+                                    update_quadrature_points);
+      unsigned int             c = 0;
       for (const auto &cell : immersed_dh.active_cell_iterators())
         {
-          // Reinitialize the cell and the fe_values
-          fe_v.reinit(cell);
-          const std::vector<Point<spacedim>> &x_points =
-            fe_v.get_quadrature_points();
-
-          // Copy the points to the vector
-          std::copy(x_points.begin(),
-                    x_points.end(),
-                    all_points.begin() + c * n_q_points);
+          bool use_cell = true;
+          if (optimize_mpi)
+            {
+              const auto &bbox = cell->bounding_box();
+              std::vector<std::pair<
+                BoundingBox<spacedim>,
+                typename Triangulation<dim0, spacedim>::active_cell_iterator>>
+                out_vals;
+              cache.get_cell_bounding_boxes_rtree().query(
+                boost::geometry::index::intersects(bbox),
+                std::back_inserter(out_vals));
+              use_cell = false;
+              // Check if the cell overlaps a locally owned cell
+              for (const auto &bbox_it : out_vals)
+                if (bbox_it.second->is_locally_owned())
+                  {
+                    use_cell = true;
+                    used_cells.emplace_back(c);
+                    break;
+                  }
+            }
+
+          if (use_cell)
+            {
+              // Reinitialize the cell and the fe_values
+              fe_v.reinit(cell);
+              const std::vector<Point<spacedim>> &x_points =
+                fe_v.get_quadrature_points();
+
+              // Insert the points to the vector
+              all_points.insert(all_points.end(),
+                                x_points.begin(),
+                                x_points.end());
+            }
```

Comment:
I think this is identical to the logic used in the first function. Could we put this in some internal helper function?

Reply:
I didn't think about this...now everything is packed into an internal function ( NonMatching::internal::compute_used_cells ). Thank you for the review!